### PR TITLE
Remove NN Builder API

### PR DESCRIPTION
**Pull request by l-bat**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

Xtest_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by dkurt** [comment]
@l-bat, I think that we also should remove these releases support, right?
```
#define INF_ENGINE_RELEASE_2018R5 2018050000
#define INF_ENGINE_RELEASE_2019R1 2019010000
#define INF_ENGINE_RELEASE_2019R2 2019020000
#define INF_ENGINE_RELEASE_2019R3 2019030000
```

File: modules/dnn/src/layers/fully_connected_layer.cpp [hunk 1]
```diff
@@ -43,7 +43,8 @@
 #include "../precomp.hpp"
 #include "layers_common.hpp"
 #include "../op_halide.hpp"
-#include "../op_inf_engine.hpp"
+// #include "../op_inf_engine.hpp"
```
Comment (dkurt) [inline modules/dnn/src/layers/fully_connected_layer.cpp]: Please remove commented code

File: modules/dnn/test/test_backends.cpp [hunk 2]
```diff
@@ -78,17 +78,6 @@ class DNNTestNetwork : public DNNTestLayer
     {
         if (outputLayer == "detection_out")
         {
-            if (backend == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019)
```
Comment (dkurt) [inline modules/dnn/test/test_backends.cpp]: Can you please check why nGraph path works correctly this way? I thought outputs from both APIs should be the same.

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 3]
```diff
@@ -284,9 +279,7 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
          */
         virtual Ptr<BackendNode> initHalide(const std::vector<Ptr<BackendWrapper> > &inputs);
 
-        virtual Ptr<BackendNode> initInfEngine(const std::vector<Ptr<BackendWrapper> > &inputs);
-
-        virtual Ptr<BackendNode> initNgraph(const std::vector<Ptr<BackendWrapper> > &inputs, const std::vector<Ptr<BackendNode> >& nodes);
+        virtual Ptr<BackendNode> initNgraph(const std::vector<Ptr<BackendNode> >& nodes);
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: What do you think, it's better to rename it to initial `initInfEngine`? But let's discuss before change :)

File: modules/dnn/src/dnn.cpp [hunk 4]
```diff
@@ -588,8 +535,7 @@ struct DataLayer : public Layer
 
     virtual bool supportBackend(int backendId) CV_OVERRIDE
     {
-        return backendId == DNN_BACKEND_OPENCV ||
-               (backendId == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019 && inputsData.size() == 1);
+        return backendId == DNN_BACKEND_OPENCV;
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: We need to support this layer with nGraph as for me

File: modules/dnn/src/dnn.cpp [hunk 5]
```diff
@@ -1471,20 +1366,12 @@ struct Net::Impl
...
-#else
-            CV_Assert(false && "This OpenCV version is built without Inference Engine NN Builder API support");
-#endif
-        }
-        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+        else if (preferableBackend == DNN_BACKEND_INFERENCE_ENGINE)
         {
-#ifdef HAVE_DNN_NGRAPH
+#ifdef HAVE_INF_ENGINE
             initNgraphBackend(blobsToKeep_);
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: Maybe let's rename `initNgraphBackend` to `initInfEngineBackend`

File: modules/dnn/src/dnn.cpp (hunk 4 above)
Comment (l-bat) [reply modules/dnn/src/dnn.cpp]: Ddo we need `inputData.size () == 1`condition?

File: modules/dnn/test/test_backends.cpp (hunk 2 above)
Comment (l-bat) [reply modules/dnn/test/test_backends.cpp]: It strange, but nGraph works in both cases

File: modules/dnn/src/dnn.cpp (hunk 4 above)
Comment (dkurt) [reply modules/dnn/src/dnn.cpp]: At least match DNN Builder implementation

**Comment by asmorkalov** [comment]
@l-bat @dkurt Do you plan to work on the patch in mean time? There are merge conflicts.

**Comment by alalek** [comment]
Lets hold this activity for now.

**Comment by JulienMaille** [comment]
@alalek is there a way to compile current `master` with legacy IE disabled (and avoid linking with `inference_engine_legacy.dll`)?
Thanks

**Pull request closed without merging**